File: visualization/src/types/unified.ts
// Unified Knowledge Base Types
// These types mirror the Python unified models for the resolved/merged entities

export interface EntityOccurrence {
  juan_index: number;
  segment_index: number;
  chunk_index: number;
  sentence_indexes: number[];
  original_description: string;
  source_sentence: string;
}

export interface UnifiedRole {
  id: string;
  canonical_name: string;
  all_names: string[];
  description: string;
  original_descriptions: string[];
  powers: string[];
  primary_power: string | null;
  first_appearance_juan: number;
  last_appearance_juan: number;
  active_period: string | null;
  occurrences: EntityOccurrence[];
  total_mentions: number;
  juans_appeared: number[];
  related_entities: string[];
  created_at: string;
  updated_at: string;
}

export interface UnifiedLocation {
  id: string;
  canonical_name: string;
  all_names: string[];
  location_type: string;
  description: string;
  modern_name: string;
  coordinates: [number, number] | null;
  associated_entities: string[];
  associated_events: string[];
  occurrences: EntityOccurrence[];
  total_mentions: number;
  juans_appeared: number[];
  created_at: string;
  updated_at: string;
}

export interface UnifiedEvent {
  id: string;
  name: string;
  time: string | null;
  time_start: number | null;
  time_end: number | null;
  location: string | null;
  participants: string[];
  description: string;
  background: string;
  significance: string;
  source_juans: number[];
  source_segments: string[];
  action_count: number;
  created_at: string;
  updated_at: string;
}

export interface UnifiedRelation {
  id: string;
  from_entity: string;
  to_entity: string;
  action_types: string[];
  primary_action: string;
  interaction_count: number;
  first_interaction_time: string | null;
  last_interaction_time: string | null;
  contexts: string[];
  source_juans: number[];
  created_at: string;
  updated_at: string;
}

export interface UnifiedKnowledgeBase {
  roles: Record<string, UnifiedRole>;
  locations: Record<string, UnifiedLocation>;
  events: Record<string, UnifiedEvent>;
  relations: Record<string, UnifiedRelation>;
  
  // Indexes
  name_to_role_id: Record<string, string>;
  name_to_location_id: Record<string, string>;
  power_to_roles: Record<string, string[]>;
  juan_to_roles: Record<number, string[]>;
  juan_to_events: Record<number, string[]>;
  
  // Statistics
  total_roles: number;
  total_locations: number;
  total_events: number;
  total_relations: number;
  juans_processed: number[];
  last_updated: string;
}

// Adapted types for visualization (compatible with existing components)
export interface RoleNodeUnified {
  id: string;
  name: string;
  power: string | null;
  description: string;
  appearances: number;
  juans: number[];
  aliases: string[];
  relatedEntities: string[];
}

export interface RoleLinkUnified {
  source: string;
  target: string;
  action: string;
  weight: number;
  time: string | null;
  actionTypes: string[];
  contexts: string[];
  sourceJuans: number[];
}

export interface TimelineEventUnified {
  id: string;
  name: string;
  time: string | null;
  timeNumeric: number | null;
  location: string | null;
  participants: string[];
  description: string;
  juan_index: number;
  type: 'event';
  significance: string;
  background: string;
}

export interface PowerDistributionUnified {
  power: string;
  count: number;
  roles: string[];
}

File: model/unified.py
"""
Unified Entity Models: Canonical representations that merge multiple occurrences.

These models aggregate information from multiple extractions across 294 卷,
providing a single source of truth for each entity, location, and event.
"""

from typing import List, Dict, Set, Optional
from pydantic import BaseModel, Field, computed_field
from datetime import datetime


class EntityOccurrence(BaseModel):
    """Record of where an entity was mentioned."""
    juan_index: int
    segment_index: int
    chunk_index: int
    sentence_indexes: List[int] = Field(default_factory=list)
    original_description: str = ""
    source_sentence: str = ""


class UnifiedRole(BaseModel):
    """
    Canonical entity that merges all occurrences of the same person/organization.
    
    This is the "resolved" version of Role, aggregating information from
    multiple extractions across the entire 资治通鉴.
    """
    
    # Unique identifier (canonical name)
    id: str = Field(description="Unique identifier, typically the most common name")
    
    # Names and aliases
    canonical_name: str = Field(description="The authoritative name for this entity")
    all_names: Set[str] = Field(
        default_factory=set,
        description="All names and aliases found across all occurrences"
    )
    
    # Merged description (best quality)
    description: str = Field(
        default="",
        description="Best description, selected from the most detailed occurrence"
    )
    
    # All original descriptions for reference
    original_descriptions: List[str] = Field(
        default_factory=list,
        description="All unique original descriptions from the book"
    )
    
    # Power/faction affiliations (may change over time)
    powers: List[str] = Field(
        default_factory=list,
        description="All power affiliations, in chronological order if known"
    )
    primary_power: Optional[str] = Field(
        default=None,
        description="Most common or significant power affiliation"
    )
    
    # Temporal information
    first_appearance_juan: int = Field(default=0)
    last_appearance_juan: int = Field(default=0)
    active_period: Optional[str] = Field(
        default=None,
        description="Time period when this entity was active, e.g., '前453年-前403年'"
    )
    
    # Occurrence tracking
    occurrences: List[EntityOccurrence] = Field(
        default_factory=list,
        description="All places where this entity was mentioned"
    )
    
    # Statistics
    total_mentions: int = Field(default=0)
    juans_appeared: Set[int] = Field(default_factory=set)
    
    # Relationships summary
    related_entities: Set[str] = Field(
        default_factory=set,
        description="Other entities this one has relationships with"
    )
    
    # Metadata
    created_at: str = Field(default_factory=lambda: datetime.now().isoformat())
    updated_at: str = Field(default_factory=lambda: datetime.now().isoformat())
    
    @computed_field
    @property
    def appearance_count(self) -> int:
        return len(self.occurrences)
    
    @computed_field
    @property
    def juan_span(self) -> int:
        """How many 卷s this entity spans."""
        return len(self.juans_appeared)
    
    class Config:
        # Allow set serialization
        json_encoders = {
            set: list
        }


class UnifiedLocation(BaseModel):
    """
    Canonical location that merges all occurrences of the same place.
    """
    
    id: str
    canonical_name: str
    all_names: Set[str] = Field(default_factory=set)
    
    location_type: str = Field(default="", description="国家、城市、地区、山川等")
    description: str = ""
    modern_name: str = ""
    coordinates: Optional[tuple[float, float]] = None
    
    # Related entities that were associated with this location
    associated_entities: Set[str] = Field(default_factory=set)
    
    # Events that occurred at this location
    associated_events: List[str] = Field(default_factory=list)
    
    occurrences: List[EntityOccurrence] = Field(default_factory=list)
    total_mentions: int = Field(default=0)
    juans_appeared: Set[int] = Field(default_factory=set)
    
    created_at: str = Field(default_factory=lambda: datetime.now().isoformat())
    updated_at: str = Field(default_factory=lambda: datetime.now().isoformat())
    
    class Config:
        json_encoders = {set: list}


class UnifiedEvent(BaseModel):
    """
    Canonical event that may span multiple segments/chunks.
    """
    
    id: str
    name: str
    
    time: Optional[str] = None
    time_start: Optional[int] = Field(default=None, description="Numeric year, negative for BC")
    time_end: Optional[int] = Field(default=None, description="Numeric year, negative for BC")
    
    location: Optional[str] = None
    
    # All participants across all mentions
    participants: Set[str] = Field(default_factory=set)
    
    description: str = ""
    background: str = ""
    significance: str = ""
    
    # Source tracking
    source_juans: Set[int] = Field(default_factory=set)
    source_segments: List[str] = Field(
        default_factory=list,
        description="List of 'juan-segment' keys"
    )
    
    # Related actions that compose this event
    action_count: int = Field(default=0)
    
    created_at: str = Field(default_factory=lambda: datetime.now().isoformat())
    updated_at: str = Field(default_factory=lambda: datetime.now().isoformat())
    
    class Config:
        json_encoders = {set: list}


class UnifiedRelation(BaseModel):
    """
    Aggregated relationship between two entities.
    
    Combines all interactions between a pair of entities across the entire text.
    """
    
    id: str = Field(description="Format: 'from_entity->to_entity'")
    
    from_entity: str
    to_entity: str
    
    # All action types between these entities
    action_types: List[str] = Field(default_factory=list)
    primary_action: str = Field(default="", description="Most common action type")
    
    # Interaction count
    interaction_count: int = Field(default=0)
    
    # Time span of interactions
    first_interaction_time: Optional[str] = None
    last_interaction_time: Optional[str] = None
    
    # Contexts from each interaction
    contexts: List[str] = Field(default_factory=list)
    
    # Source tracking
    source_juans: Set[int] = Field(default_factory=set)
    
    created_at: str = Field(default_factory=lambda: datetime.now().isoformat())
    updated_at: str = Field(default_factory=lambda: datetime.now().isoformat())
    
    class Config:
        json_encoders = {set: list}


class UnifiedKnowledgeBase(BaseModel):
    """
    The complete unified knowledge base for 资治通鉴.
    
    This is the "gold standard" data structure that visualization
    and search should use.
    """
    
    # Entity registries
    roles: Dict[str, UnifiedRole] = Field(default_factory=dict)
    locations: Dict[str, UnifiedLocation] = Field(default_factory=dict)
    events: Dict[str, UnifiedEvent] = Field(default_factory=dict)
    relations: Dict[str, UnifiedRelation] = Field(default_factory=dict)
    
    # Name resolution index: maps any name/alias to canonical ID
    name_to_role_id: Dict[str, str] = Field(default_factory=dict)
    name_to_location_id: Dict[str, str] = Field(default_factory=dict)
    
    # Power/faction index
    power_to_roles: Dict[str, List[str]] = Field(default_factory=dict)
    
    # Temporal index (juan -> entities active in that juan)
    juan_to_roles: Dict[int, List[str]] = Field(default_factory=dict)
    juan_to_events: Dict[int, List[str]] = Field(default_factory=dict)
    
    # Statistics
    total_roles: int = Field(default=0)
    total_locations: int = Field(default=0)
    total_events: int = Field(default=0)
    total_relations: int = Field(default=0)
    
    juans_processed: List[int] = Field(default_factory=list)
    last_updated: str = Field(default_factory=lambda: datetime.now().isoformat())
    
    class Config:
        json_encoders = {set: list}
